src/pond.rs

```diff
@@ -0,0 +1,282 @@
+use std::path::Component;
+use std::path::Path;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use crd::CRDSpec;
+use anyhow::{Context,Result,anyhow};
+
+use arrow::datatypes::{DataType, Field, Fields, FieldRef};
+use serde::{Serialize, Deserialize};
+
+//use std::ffi::OsStr;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct PondResource {
+    kind: String,
+    name: String,
+    api_version: String,
+    uuid: Uuid,
+    metadata: Option<BTreeMap<String, String>>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct UniqueSpec<T> {
+    uuid: Uuid,
+
+    #[serde(flatten)]
+    spec: T,
+}
+
+#[derive(Debug)]
+pub struct Pond {
+    pub dirs: BTreeMap<PathBuf, dir::Directory>,
+    pub resources: Vec<PondResource>,
+}
+
+fn resource_fields() -> Vec<FieldRef> {
+    vec![
+        Arc::new(Field::new("kind", DataType::Utf8, false)),
+        Arc::new(Field::new("apiVersion", DataType::Utf8, false)),
+        Arc::new(Field::new("name", DataType::Utf8, false)),
+        Arc::new(Field::new("uuid", DataType::Utf8, false)),
+        Arc::new(Field::new("metadata",
+			    DataType::Map(
+				Arc::new(
+				    Field::new("entries",
+					       DataType::Struct(Fields::from(vec![
+						   Field::new("key", DataType::Utf8, false),
+						   Field::new("value", DataType::Utf8, false),
+					       ])),
+					       false,
+				    )),
+				false),
+			    true)),
+    ]
+}
+
+fn hydrovu_fields() -> Vec<FieldRef> {
+    vec![
+        Arc::new(Field::new("uuid", DataType::Utf8, false)),
+        Arc::new(Field::new("key", DataType::Utf8, false)),
+        Arc::new(Field::new("secret", DataType::Utf8, false)),
+    ]
+}
+
+pub fn find_pond() -> Result<Option<PathBuf>> {
+    let path = std::env::current_dir().with_context(|| "could not get working directory")?;
+	
+    find_recursive(path.as_path())
+}
+
+fn find_recursive<P: AsRef<Path>>(path: P) -> Result<Option<PathBuf>> {
+    let mut ppath = path.as_ref().to_path_buf();
+    ppath.push(".pond");
+    let path = ppath.as_path();
+    if path.is_dir() {
+	return Ok(Some(path.to_path_buf()));
+    }
+    ppath.pop();
+    if let Some(parent) = ppath.parent() {
+	return find_recursive(parent);
+    }
+    Ok(None)
+}
+
+pub fn init() -> Result<()> {
+    let has = find_pond()?;
+    if let Some(path) = has {
+	return Err(anyhow!("pond exists! {:?}", path));
+    }
+
+    let mut directory = dir::create_dir(".pond")?;
+
+    let empty: Vec<PondResource> = vec![];
+    directory.write_file("pond", &empty, resource_fields().as_slice())?;
+
+    directory.close_dir()?;
+
+    Ok(())
+}
+
+pub fn open() -> Result<Pond> {
+    let loc = find_pond()?;
+    if let None = loc {
+	return Err(anyhow!("pond does not exist"))
+    }
+    let path = loc.unwrap().clone();
+    let root = dir::open_dir(&path)?;
+    let pond_path = root.current_path_of("pond")?;
+    let mut dirs: BTreeMap<PathBuf, dir::Directory> = BTreeMap::new();
```

Comment:
The compiler is probably able to infer the entire type signature of this variable.